libclamav_rust/src/fuzzy_hash.rs

```diff
@@ -0,0 +1,544 @@
+    hashmap.get(&hash)
+}
+
+/// C interface for fuzzy_hash_load_subsignature().
+/// Handles all the unsafe ffi stuff.
+#[export_name = "fuzzy_hash_load_subsignature"]
+pub extern "C" fn _fuzzy_hash_load_subsignature(
+    matcher: *const sys::cli_matcher,
+    hexsig: *const c_char,
+    lsig_id: u32,
+    subsig_id: u32,
+) -> sys::cl_error_t {
+    let hexsig = validate_str_param!(hexsig);
+
+    let p = unsafe { (*matcher).fuzzy_hashmap };
+    let mut hashmap = unsafe {
+        ManuallyDrop::new(Box::from_raw(
+            p as *mut HashMap<FuzzyHash, Vec<FuzzyHashMeta>>,
+        ))
+    };
+
+    let result: sys::cl_error_t =
+        match fuzzy_hash_load_subsignature(&mut hashmap, hexsig, lsig_id, subsig_id) {
+            Ok(_) => sys::cl_error_t_CL_SUCCESS,
+            Err(error) => {
+                error!("Error when loading fuzzy hash logical subsignature:");
+                error!("  {}", hexsig);
+
+                let result = match error {
+                    FuzzyHashError::Format => {
+                        error!("Invalid signature format.");
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    FuzzyHashError::UnknownAlgorithm(algo) => {
+                        error!("Unknown hashing algorithm: {}", algo);
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    FuzzyHashError::FormatHammingDistance(dist) => {
+                        error!("Invalid hamming distance: {}", dist);
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    FuzzyHashError::FormatHashBytes(hash) => {
+                        error!("Invalid bytes for hash: {}", hash);
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    _ => {
+                        error!("Unexpected error");
+                        sys::cl_error_t_CL_ERROR
+                    }
+                };
+
+                error!("Expected format: algorithm#hash[#hammingdistance]");
+                error!("  where");
+                error!("   - algorithm:       Must be 'fuzzy_img'");
+                error!("   - hash:            Must be an 8-byte hex string");
+                error!("   - hammingdistance: (optional) Must be an unsigned integer");
+
+                result
+            }
+        };
+
+    result
+}
+
+/// Load a fuzzy hash subsignature
+/// Parse a fuzzy hash logical sig subsignature.
+/// Add the fuzzy hash to the matcher so it can be matched.
+pub fn fuzzy_hash_load_subsignature(
+    hashmap: &mut Box<HashMap<FuzzyHash, Vec<FuzzyHashMeta>>>,
+    hexsig: &str,
+    lsig_id: u32,
+    subsig_id: u32,
+) -> Result<(), FuzzyHashError> {
+    let mut hexsig_split = hexsig.split('#');
+
+    let algorithm = match hexsig_split.next() {
+        Some(x) => x,
+        None => return Err(FuzzyHashError::Format),
+    };
+
+    let hash = match hexsig_split.next() {
```

Comment:
You can catch two errors at once here if you use `hex::decode_to_slice`.  See `util.rs` in the signature validator code for an example.  It's a touch more efficient since it doesn't require heap allocation.

Reply:
Discussed out-of-band. I updated the relevant code to use `hex::decode_to_slice` (below)